File: 8-string-to-integer-atoi/8. String to Integer (atoi).java
class Solution {
    public int myAtoi(String s) {
        s = s.trim(); // Remove leading and trailing whitespaces
        if (s.isEmpty()) {
            return 0;
        }

        int i = 0;
        int sign = 1; // to handle the sign of the number
        int result = 0;

        // Check for optional sign
        if (s.charAt(i) == '+' || s.charAt(i) == '-') {
            if (s.charAt(i) == '-') {
                sign = -1;
            }
            i++;
        }

        // Process numerical digits and form the number
        while (i < s.length()) {
            char ch = s.charAt(i);
            if (ch < '0' || ch > '9') {
                break; // break if the current character is not a digit
            }

            // Check for overflow and underflow
            if (result > (Integer.MAX_VALUE - (ch - '0')) / 10) {
                return (sign == 1) ? Integer.MAX_VALUE : Integer.MIN_VALUE;
            }

            result = result * 10 + (ch - '0');
            i++;
        }

        return result * sign;
    }
}

File: 8-string-to-integer-atoi/8. String to Integer (atoi).cpp
class Solution {
public:
    int myAtoi(string s) {
    int res=0;
    int i=0;
    int sign=1;
	
    while(i<s.size()&&s[i]==' ')i++;  
    if(s[i]=='-'||s[i]=='+')         
    {
        sign=s[i]=='-'?-1:1;
        i++;
    }
   
    while(i<s.length()&&(s[i]>='0'&&s[i]<='9'))  
    {
        int digit=(s[i]-'0')*sign;
        if(sign==1 && (res>INT_MAX/10 || (res==INT_MAX/10 && digit>INT_MAX%10))) return INT_MAX;  
        if(sign==-1 &&(res<INT_MIN/10 || (res==INT_MIN/10 && digit<INT_MIN%10))) return INT_MIN;  
        res=res*10+digit;
        i++;
    }

return res;
    }
};